Cold start on android took 5-10 seconds longer than before the library update

In library version 8.1.3, where the setUserAuthenticationParameters method was changed using KeyProperties.AUTH_BIOMETRIC_STRONG the application started taking much longer to run

We are also experiencing this, however rolling back to 8.1.2 we see this exception using logcat:

```
StrongBox security storage is not available.
android.security.keystore.StrongBoxUnavailableException: Failed to generated key pair.
	at android.security.keystore2.AndroidKeyStoreKeyPairGeneratorSpi.generateKeyPair(AndroidKeyStoreKeyPairGeneratorSpi.java:662)
	at java.security.KeyPairGenerator$Delegate.generateKeyPair(KeyPairGenerator.java:746)
	at ue.i.r(SourceFile:12)
	at ue.b.F(SourceFile:20)
	at ue.b.E(SourceFile:2)
	at ue.b.s(SourceFile:26)
	at com.oblador.keychain.KeychainModule.internalWarmingBestCipher(SourceFile:48)
	at com.oblador.keychain.KeychainModule.a(SourceFile:1)
	at com.oblador.keychain.b.run(SourceFile:1)
	at java.lang.Thread.run(Thread.java:1012)
Caused by: android.security.KeyStoreException: Requested security level (likely Strongbox) is not available. (internal Keystore
Caused by:
    Error::Km(r#HARDWARE_TYPE_UNAVAILABLE)) (public error code: 12 internal Keystore code: -68)
	at android.security.KeyStore2.getKeyStoreException(KeyStore2.java:435)
	at android.security.KeyStore2.handleRemoteExceptionWithRetry(KeyStore2.java:121)
	at android.security.KeyStore2.getSecurityLevel(KeyStore2.java:284)
	at android.security.keystore2.AndroidKeyStoreKeyPairGeneratorSpi.generateKeyPair(AndroidKeyStoreKeyPairGeneratorSpi.java:649)
```

It seems that it still works, have tested on a Pixel 7a and a Samsung Galaxy A54.

But is it OK to go into production with exception?

```
keystore2::error: In get_security_level: No such security level.
                                                                                                    
                                                                                                    Caused by:
                                                                                                        Error::Km(ErrorCode(-68))
2024-07-09 17:29:33.395  7187-7281  c                       pid-7187                             W  StrongBox security storage is not available.
                                                                                                    android.security.keystore.StrongBoxUnavailableException: Failed to generate key
                                                                                                    	at android.security.keystore2.AndroidKeyStoreKeyGeneratorSpi.engineGenerateKey(AndroidKeyStoreKeyGeneratorSpi.java:411)
                                                                                                    	at javax.crypto.KeyGenerator.generateKey(KeyGenerator.java:612)
                                                                                                    	at mh.i.r(CipherStorageKeystoreAesCbc.java:14)
                                                                                                    	at mh.c.F(CipherStorageBase.java:20)
                                                                                                    	at mh.c.E(CipherStorageBase.java:2)
                                                                                                    	at mh.c.s(CipherStorageBase.java:27)
                                                                                                    	at com.oblador.keychain.KeychainModule.internalWarmingBestCipher(KeychainModule.java:49)
                                                                                                    	at com.oblador.keychain.KeychainModule.a(KeychainModule.java:1)
                                                                                                    	at com.oblador.keychain.b.run(R8$$SyntheticClass:3)
                                                                                                    	at java.lang.Thread.run(Thread.java:1012)
                                                                                                    Caused by: android.security.KeyStoreException: No StrongBox available (public error code: 12 internal Keystore code: -68)
```

I had this problem, everything worked fine 2 weeks ago, I'm trying to figure out what went wrong
react-native: 0.72.4
react-native-keychain: 8.1.3

--- a/.github/workflows/e2e_tests.yaml
+++ b/.github/workflows/e2e_tests.yaml
@@ -30,8 +30,8 @@ jobs:
       - name: Setup java
         uses: actions/setup-java@v4
         with:
-          distribution: 'temurin' # See 'Supported distributions' for available options
-          java-version: '19'
+          distribution: 'temurin'
+          java-version: '17'
 
       - name: Cache Gradle
         uses: actions/cache@v4
@@ -99,8 +99,8 @@ jobs:
       - name: Setup java
         uses: actions/setup-java@v4
         with:
-          distribution: 'temurin' # See 'Supported distributions' for available options
-          java-version: '19'
+          distribution: 'temurin'
+          java-version: '17'
 
       - uses: actions/download-artifact@v4
         with:

--- a/.github/workflows/gradle_build.yaml
+++ b/.github/workflows/gradle_build.yaml
@@ -31,7 +31,7 @@ jobs:
         uses: actions/setup-java@v4
         with:
           distribution: 'temurin'
-          java-version: '19'
+          java-version: '17'
 
       - name: Cache Gradle
         uses: actions/cache@v4

--- a/.github/workflows/gradle_build.yaml
+++ b/.github/workflows/gradle_build.yaml
@@ -1,0 +1,57 @@
+name: Gradle Build
+
+on:
+  push:
+    branches:
+      - master
+  pull_request:
+    branches:
+      - master
+
+jobs:
+  build:
+    name: Build react-native-keychain
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v4
+
+      - name: Setup Node.js
+        uses: actions/setup-node@v4
+        with:
+          node-version-file: .nvmrc
+
+      - name: Enable corepack
+        run: corepack enable
+
+      - name: Install dependencies
+        run: yarn install --immutable
+
+      - name: Setup java
+        uses: actions/setup-java@v4
+        with:
+          distribution: 'temurin'
+          java-version: '19'
+
+      - name: Cache Gradle
+        uses: actions/cache@v4
+        with:
+          path: |
+            ~/.gradle/wrapper
+            ~/.gradle/caches
+          key: ${{ runner.os }}-gradleLibrary-${{ hashFiles('android/gradle/wrapper/gradle-wrapper.properties') }}
+          restore-keys: |
+            ${{ runner.os }}-gradleLibrary-
+
+      - name: Downloading Gradle Dependencies
+        run: |
+          cd android
+          chmod +x gradlew
+          ./gradlew --max-workers 2 dependencies
+
+      - name: Build Library
+        env:
+          JAVA_OPTS: '-XX:MaxHeapSize=6g'
+        working-directory: ./android
+        run: |
+          ./gradlew build

--- a/android/build.gradle
+++ b/android/build.gradle
@@ -2,11 +2,8 @@ buildscript {
   // Buildscript is evaluated before everything else so we can't use safeExtGet
   def kotlin_version = rootProject.ext.has('kotlinVersion') ? rootProject.ext.get('kotlinVersion') : project.properties['RNKeychain_kotlinVersion']
   repositories {
-    maven {
-      url "https://plugins.gradle.org/m2/"
-    }
-    mavenCentral()
     google()
+    mavenCentral()
   }
   dependencies {
     classpath "com.android.tools.build:gradle:8.3.1"
@@ -15,7 +12,7 @@ buildscript {
 }
 
 def safeExtGet(prop, fallback) {
-  rootProject.ext.has(prop) ? rootProject.ext.get(prop) : fallback
+  return rootProject.ext.has(prop) ? rootProject.ext.get(prop) : fallback
 }
 
 def isNewArchitectureEnabled() {
@@ -30,37 +27,31 @@ if (isNewArchitectureEnabled()) {
   apply plugin: "com.facebook.react"
 }
 
-repositories {
-  google()
-  mavenCentral()
-}
-
 android {
-  // Conditional for compatibility with AGP <4.2.
-  if (project.android.hasProperty("namespace")) {
+  def agpVersion = com.android.Version.ANDROID_GRADLE_PLUGIN_VERSION
+  if (agpVersion.tokenize('.')[0].toInteger() >= 7) {
     namespace = "com.oblador.keychain"
+    buildFeatures {
+      buildConfig true
+    }
   }
-  compileSdkVersion safeExtGet('compileSdkVersion', project.properties['RNKeychain_compileSdkVersion'])
+  compileSdkVersion safeExtGet('compileSdkVersion', project.properties['RNKeychain_compileSdkVersion'].toInteger())
 
   defaultConfig {
-    minSdkVersion safeExtGet('minSdkVersion', project.properties['RNKeychain_minSdkVersion'])
-    compileSdkVersion safeExtGet('compileSdkVersion', project.properties['RNKeychain_compileSdkVersion'])
-    targetSdkVersion safeExtGet('targetSdkVersion', project.properties['RNKeychain_targetSdkVersion'])
+    minSdkVersion safeExtGet('minSdkVersion', project.properties['RNKeychain_minSdkVersion'].toInteger())
+    targetSdkVersion safeExtGet('targetSdkVersion', project.properties['RNKeychain_targetSdkVersion'].toInteger())
     buildConfigField "boolean", "IS_NEW_ARCHITECTURE_ENABLED", isNewArchitectureEnabled().toString()
   }
 
   compileOptions {
-    sourceCompatibility JavaVersion.VERSION_1_8
-    targetCompatibility JavaVersion.VERSION_1_8
-  }
-
-  buildFeatures {
-    buildConfig true
+    sourceCompatibility JavaVersion.VERSION_17
+    targetCompatibility JavaVersion.VERSION_17
   }
 }
 
 repositories {
   mavenCentral()
+  google()
 }
 
 def kotlin_version = safeExtGet('kotlinVersion', project.properties['RNKeychain_kotlinVersion'])

--- a/android/gradle.properties
+++ b/android/gradle.properties
@@ -11,4 +11,4 @@ android.enableJetifier=true
 RNKeychain_kotlinVersion=1.8.0
 RNKeychain_compileSdkVersion=34
 RNKeychain_targetSdkVersion=34
-RNKeychain_minSdkVersion=23
+RNKeychain_minSdkVersion=21

--- a/android/src/main/AndroidManifest.xml
+++ b/android/src/main/AndroidManifest.xml
@@ -1,8 +1,10 @@
 <manifest
   xmlns:android="http://schemas.android.com/apk/res/android"
   package="com.oblador.keychain"
+  xmlns:tools="http://schemas.android.com/tools"
   >
 
+  <uses-sdk tools:overrideLibrary="com.facebook.react" />
   <uses-permission android:name="android.permission.USE_BIOMETRIC" />
   <uses-permission android:name="android.permission.USE_FINGERPRINT" />
 

--- a/android/src/main/java/com/oblador/keychain/KeychainModule.kt
+++ b/android/src/main/java/com/oblador/keychain/KeychainModule.kt
@@ -138,11 +138,11 @@ class KeychainModule(reactContext: ReactApplicationContext) :
   init {
     prefsStorage = PrefsStorage(reactContext)
     addCipherStorageToMap(CipherStorageFacebookConceal(reactContext))
-    addCipherStorageToMap(CipherStorageKeystoreAesCbc())
+    addCipherStorageToMap(CipherStorageKeystoreAesCbc(reactContext))
 
     // we have a references to newer api that will fail load of app classes in old androids OS
     if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
-      addCipherStorageToMap(CipherStorageKeystoreRsaEcb())
+      addCipherStorageToMap(CipherStorageKeystoreRsaEcb(reactContext))
     }
   }
 

--- a/android/src/main/java/com/oblador/keychain/cipherStorage/CipherStorageBase.kt
+++ b/android/src/main/java/com/oblador/keychain/cipherStorage/CipherStorageBase.kt
@@ -6,7 +6,6 @@ import android.os.Build
 import android.security.keystore.KeyGenParameterSpec
 import android.security.keystore.KeyInfo
 import android.util.Log
-import androidx.annotation.RequiresApi
 import androidx.annotation.VisibleForTesting
 import com.oblador.keychain.SecurityLevel
 import com.oblador.keychain.cipherStorage.CipherStorageBase.DecryptBytesHandler

--- a/android/src/main/java/com/oblador/keychain/cipherStorage/CipherStorageBase.kt
+++ b/android/src/main/java/com/oblador/keychain/cipherStorage/CipherStorageBase.kt
@@ -1,11 +1,16 @@
 package com.oblador.keychain.cipherStorage
 
+import android.content.Context
+import android.content.pm.PackageManager
 import android.os.Build
 import android.security.keystore.KeyGenParameterSpec
 import android.security.keystore.KeyInfo
 import android.util.Log
+import androidx.annotation.RequiresApi
 import androidx.annotation.VisibleForTesting
 import com.oblador.keychain.SecurityLevel
+import com.oblador.keychain.cipherStorage.CipherStorageBase.DecryptBytesHandler
+import com.oblador.keychain.cipherStorage.CipherStorageBase.EncryptStringHandler
 import com.oblador.keychain.exceptions.CryptoFailedException
 import com.oblador.keychain.exceptions.KeyStoreAccessException
 import java.io.ByteArrayInputStream
@@ -23,7 +28,6 @@ import java.security.NoSuchAlgorithmException
 import java.security.ProviderException
 import java.security.UnrecoverableKeyException
 import java.util.Collections
-import java.util.HashSet
 import java.util.concurrent.atomic.AtomicBoolean
 import java.util.concurrent.atomic.AtomicInteger
 import javax.crypto.Cipher
@@ -32,8 +36,9 @@ import javax.crypto.CipherOutputStream
 import javax.crypto.NoSuchPaddingException
 import javax.crypto.spec.IvParameterSpec
 
+
 @Suppress("unused", "MemberVisibilityCanBePrivate", "UnusedReturnValue")
-abstract class CipherStorageBase : CipherStorage {
+abstract class CipherStorageBase(protected val applicationContext: Context) : CipherStorage {
   // region Constants
   /** Logging tag. */
   protected val LOG_TAG = CipherStorageBase::class.java.simpleName
@@ -82,22 +87,26 @@ abstract class CipherStorageBase : CipherStorage {
   /** Guard object for [isSupportsSecureHardware] field. */
   protected val _sync = Any()
 
+  /** Try to resolve support of the StrongBox feature. */
+  protected val isStrongboxAvailable: Boolean by lazy {
+    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
+      applicationContext.packageManager.hasSystemFeature(PackageManager.FEATURE_STRONGBOX_KEYSTORE)
+    } else {
+      false
+    }
+  }
+
   /** Try to resolve it only once and cache result for all future calls. */
-  @Transient protected var isSupportsSecureHardware: AtomicBoolean? = null
-
-  /** Guard for [isStrongboxAvailable] field assignment. */
-  protected val _syncStrongbox = Any()
-
-  /** Try to resolve support of the strongbox and cache result for future calls. */
-  @Transient protected var isStrongboxAvailable: AtomicBoolean? = null
+  @Transient
+  protected var isSupportsSecureHardware: AtomicBoolean? = null
 
   /** Get cached instance of cipher. Get instance operation is slow. */
-  @Transient protected var cachedCipher: Cipher? = null
+  @Transient
+  protected var cachedCipher: Cipher? = null
 
   /** Cached instance of the Keystore. */
-  @Transient protected var cachedKeyStore: KeyStore? = null
-
-  // endregion
+  @Transient
+  protected var cachedKeyStore: KeyStore? = null
 
   // region Overrides
 
@@ -133,7 +142,8 @@ abstract class CipherStorageBase : CipherStorage {
         sdk = SelfDestroyKey(TEST_KEY_ALIAS)
         val newValue = validateKeySecurityLevel(SecurityLevel.SECURE_HARDWARE, sdk.key)
         isSupportsSecureHardware!!.set(newValue)
-      } catch (ignored: Throwable) {} finally {
+      } catch (ignored: Throwable) {
+      } finally {
         sdk?.close()
       }
     }
@@ -180,12 +190,13 @@ abstract class CipherStorageBase : CipherStorage {
 
   @Throws(GeneralSecurityException::class)
   protected abstract fun getKeyGenSpecBuilder(
-      alias: String,
-      isForTesting: Boolean
+    alias: String,
+    isForTesting: Boolean
   ): KeyGenParameterSpec.Builder
 
   /** Get information about provided key. */
-  @Throws(GeneralSecurityException::class) protected abstract fun getKeyInfo(key: Key): KeyInfo
+  @Throws(GeneralSecurityException::class)
+  protected abstract fun getKeyInfo(key: Key): KeyInfo
 
   /** Try to generate key from provided specification. */
   @Throws(GeneralSecurityException::class)
@@ -219,16 +230,17 @@ abstract class CipherStorageBase : CipherStorage {
   protected fun throwIfInsufficientLevel(level: SecurityLevel) {
     if (!securityLevel().satisfiesSafetyThreshold(level)) {
       throw CryptoFailedException(
-          "Insufficient security level (wants $level; got ${securityLevel()})")
+        "Insufficient security level (wants $level; got ${securityLevel()})"
+      )
     }
   }
 
   /** Extract existing key or generate a new one. In case of problems raise exception. */
   @Throws(GeneralSecurityException::class)
   protected fun extractGeneratedKey(
-      safeAlias: String,
-      level: SecurityLevel,
-      retries: AtomicInteger
+    safeAlias: String,
+    level: SecurityLevel,
+    retries: AtomicInteger
   ): Key {
     var key: Key?
     do {
@@ -352,9 +364,9 @@ abstract class CipherStorageBase : CipherStorage {
   /** Decrypt provided bytes to a string. */
   @Throws(GeneralSecurityException::class, IOException::class)
   protected open fun decryptBytes(
-      key: Key,
-      bytes: ByteArray,
-      handler: DecryptBytesHandler?
+    key: Key,
+    bytes: ByteArray,
+    handler: DecryptBytesHandler?
   ): String {
     val cipher = getCachedInstance()
     try {
@@ -384,25 +396,19 @@ abstract class CipherStorageBase : CipherStorage {
 
     var secretKey: Key? = null
 
-    // multi-threaded usage is possible
-    synchronized(_syncStrongbox) {
-      if (isStrongboxAvailable == null || isStrongboxAvailable!!.get()) {
-        if (isStrongboxAvailable == null) isStrongboxAvailable = AtomicBoolean(false)
-
-        try {
-          secretKey = tryGenerateStrongBoxSecurityKey(alias)
-          isStrongboxAvailable!!.set(true)
-        } catch (ex: GeneralSecurityException) {
-          Log.w(LOG_TAG, "StrongBox security storage is not available.", ex)
-        } catch (ex: ProviderException) {
-          Log.w(LOG_TAG, "StrongBox security storage is not available.", ex)
-        }
+    if (isStrongboxAvailable) {
+      try {
+        secretKey = tryGenerateStrongBoxSecurityKey(alias)
+      } catch (ex: GeneralSecurityException) {
+        Log.w(LOG_TAG, "StrongBox security storage is not available.", ex)
+      } catch (ex: ProviderException) {
+        Log.w(LOG_TAG, "StrongBox security storage is not available.", ex)
       }
     }
 
     // If that is not possible, we generate the key in a regular way
     // (it still might be generated in hardware, but not in StrongBox)
-    if (secretKey == null || !isStrongboxAvailable!!.get()) {
+    if (secretKey == null || !isStrongboxAvailable) {
       try {
         secretKey = tryGenerateRegularSecurityKey(alias)
       } catch (fail: GeneralSecurityException) {
@@ -426,7 +432,8 @@ abstract class CipherStorageBase : CipherStorage {
   protected fun tryGenerateRegularSecurityKey(alias: String, isForTesting: Boolean): Key {
     if (Build.VERSION.SDK_INT < Build.VERSION_CODES.M) {
       throw KeyStoreAccessException(
-          "Regular security keystore is not supported for old API${Build.VERSION.SDK_INT}.")
+        "Regular security keystore is not supported for old API${Build.VERSION.SDK_INT}."
+      )
     }
 
     val specification = getKeyGenSpecBuilder(alias, isForTesting).build()
@@ -443,7 +450,8 @@ abstract class CipherStorageBase : CipherStorage {
   protected fun tryGenerateStrongBoxSecurityKey(alias: String, isForTesting: Boolean): Key {
     if (Build.VERSION.SDK_INT < Build.VERSION_CODES.P) {
       throw KeyStoreAccessException(
-          "Strong box security keystore is not supported for old API${Build.VERSION.SDK_INT}.")
+        "Strong box security keystore is not supported for old API${Build.VERSION.SDK_INT}."
+      )
     }
 
     val specification = getKeyGenSpecBuilder(alias, isForTesting).setIsStrongBoxBacked(true).build()
@@ -506,7 +514,7 @@ abstract class CipherStorageBase : CipherStorage {
       val iv = ByteArray(IV_LENGTH)
 
       if (IV_LENGTH >= bytes.size)
-          throw IOException("Insufficient length of input data for IV extracting.")
+        throw IOException("Insufficient length of input data for IV extracting.")
 
       System.arraycopy(bytes, 0, iv, 0, IV_LENGTH)
 

--- a/android/src/main/java/com/oblador/keychain/cipherStorage/CipherStorageFacebookConceal.kt
+++ b/android/src/main/java/com/oblador/keychain/cipherStorage/CipherStorageFacebookConceal.kt
@@ -26,7 +26,7 @@ import java.security.Key
  */
 @Suppress("unused", "MemberVisibilityCanBePrivate")
 class CipherStorageFacebookConceal(@NonNull reactContext: ReactApplicationContext) :
-    CipherStorageBase() {
+    CipherStorageBase(reactContext) {
 
   companion object {
     const val KEYCHAIN_DATA = "RN_KEYCHAIN"

--- a/android/src/main/java/com/oblador/keychain/cipherStorage/CipherStorageKeystoreAesCbc.kt
+++ b/android/src/main/java/com/oblador/keychain/cipherStorage/CipherStorageKeystoreAesCbc.kt
@@ -7,6 +7,7 @@ import android.security.keystore.KeyInfo
 import android.security.keystore.KeyProperties
 import android.util.Log
 import androidx.annotation.NonNull
+import com.facebook.react.bridge.ReactApplicationContext
 import com.oblador.keychain.KeychainModule.KnownCiphers
 import com.oblador.keychain.SecurityLevel
 import com.oblador.keychain.decryptionHandler.DecryptionResultHandler
@@ -30,7 +31,8 @@ import javax.crypto.SecretKeyFactory
  */
 @TargetApi(Build.VERSION_CODES.M)
 @Suppress("unused", "MemberVisibilityCanBePrivate")
-class CipherStorageKeystoreAesCbc : CipherStorageBase() {
+class CipherStorageKeystoreAesCbc(@NonNull reactContext: ReactApplicationContext) :
+    CipherStorageBase(reactContext) {
 
   // region Constants
   /** AES */

--- a/android/src/main/java/com/oblador/keychain/cipherStorage/CipherStorageKeystoreRsaEcb.kt
+++ b/android/src/main/java/com/oblador/keychain/cipherStorage/CipherStorageKeystoreRsaEcb.kt
@@ -7,7 +7,9 @@ import android.security.keystore.KeyInfo
 import android.security.keystore.KeyProperties
 import android.security.keystore.UserNotAuthenticatedException
 import android.util.Log
+import androidx.annotation.NonNull
 import androidx.annotation.RequiresApi
+import com.facebook.react.bridge.ReactApplicationContext
 import com.oblador.keychain.KeychainModule
 import com.oblador.keychain.SecurityLevel
 import com.oblador.keychain.decryptionHandler.DecryptionResultHandler
@@ -30,7 +32,8 @@ import javax.crypto.NoSuchPaddingException
 /** Fingerprint biometry protected storage. */
 @RequiresApi(Build.VERSION_CODES.M)
 @Suppress("unused", "WeakerAccess")
-class CipherStorageKeystoreRsaEcb : CipherStorageBase() {
+class CipherStorageKeystoreRsaEcb(@NonNull reactContext: ReactApplicationContext) :
+    CipherStorageBase(reactContext) {
 
   companion object {
     /** Selected algorithm. */

